File: LinkedList/C/E21-merge-two-sorted-lists.c
// Time:  O(m + n)
// Space: O(1)

/**
 * Definition for singly-linked list.
 * struct ListNode{
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* l1, strust ListNode* l2){
	if(l1 == NULL) return l2;
    if(l2 == NULL) return l1;
    struct ListNode *pa, *pb, *pc, *l3;
	pa = l1;
	pb = l2;

	if(pa->val <= pb->val){
		l3 = pc = pa;
		pa = pa->next;
	}else{
		l3 = pc = pb;
		pb = pb->next;
	}

	while(pa && pb){
		if(pa->val <= pb->val){
			pc->next = pa;
			pc = pa;
			pa = pa->next;
		}else{
			pc->next = pb;
			pc = pb;
			pb = pb->next;
		}
	}
	
	pc->next = pa?pa : pb;
	return l3;
}

File: LinkedList/Python3/E21-merge-two-sorted-lists.py
# Time:  O(m + n)
# Space: O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
	# My solution, iteratively	
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        l3 = pc = ListNode(0)

        while l1 and l2:
            if l1.val <= l2.val:
                pc.next = l1
                l1 = l1.next
            else:
                pc.next = l2
                l2 = l2.next
			pc = pc.next

        pc.next = l1 if l1 else l2
        return l3.next
		
	# recursively
	def mergeTwoLists(self, l1, l2):
		if not l1 or not l2:
			return l1 or l2
		if l1.val < l2.val:
			l1.next = self.mergeTwoLists(l1.next, l2)
			return l1
		else:
			l2.next = self.mergeTwoLists(l1, l2.next)
			return l2